### ISSUE TITLE
New rule for avoiding escaping selector and normal selectors

### ISSUE THREAD

**Issue by evilebottnawi** [stylelint/stylelint#3972]
> What is the problem you're trying to solve?

Input (same selectors):
```css
.u-m\00002b {}

.u-m\+ {}
```

> What solution would you like to see?

New rule what warning when you use identical selectors like above.

Maybe `no-duplicate-escape-selectors`, need discussion about name

**Comment by jeddy3** [comment]
Could this be an option for `selector-no-duplicates`, like the [recently added](https://github.com/stylelint/stylelint/pull/3936) `disallowsInList` option?

**Comment by evilebottnawi** [comment]
@jeddy3 hm, yep it is part of `no-duplicate-selectors` rule, but i think it should be by default, without option, because it is duplicate selectors

**Comment by jeddy3** [comment]
SGTM. We can treat it as a fix to the rule.

**Comment by Mouvedia** [comment]
Some browsers didn't support escapement in selectors—which was used as a css hack—hence it should be an option which would be true/enabled (remove duplicate or warn/error) by default.

https://web.archive.org/web/20051210225937/http://www.dithered.com/css_filters/css_only/escaped_selector.html
https://www.doxdesk.com/personal/posts/css/20020212-bmh.html

**Comment by jeddy3** [comment]
Thanks for digging into the history @Mouvedia . As those browsers are from over a decade ago, I think we can treat escaped selectors as the same for this purpose of this rule, and fix this as a false negatives bug. Anyone targeting these older browsers can turn the rule off.

**Comment by ybiquitous** [comment]
I intend to work on this issue, but I cannot find a good way to unescape (or normalize) selectors.

I found the following internal utility function in `postcss-selector-parser`. Would that help?
https://github.com/postcss/postcss-selector-parser/blob/v6.0.10/src/util/unesc.js

**Comment by romainmenke** [comment]
Was looking through issues related to `no-duplicate-selectors` and curious if we can resolve this differently.

The underlying issue here isn't really the duplicate selectors but the inconsistent escaping.

Can we have a rule to enforce consistent escaping?
Such a rule could promote avoiding needless escape sequences and promote the least verbose escape sequence.

Implementing the logic would be a good fit for `@csstools/css-tokenizer`.
We already have a partial implementation of this for `mutateIdent` and `mutateUnit` : https://github.com/csstools/postcss-plugins/blob/main/packages/css-tokenizer/src/util/mutations.ts

-----

If we have a rule to help with escaping then that rule could apply and autofix for `.u-m\00002b {}` -> `.u-m\+ {}` and then the selectors would be duplicates for `no-duplicate-selectors`

**Comment by Mouvedia** [comment]
> If we have a rule to help with escaping then that rule could apply and autofix for `.u-m\00002b {}` -> `.u-m\+ {}` and then the selectors would be duplicates for `no-duplicate-selectors`

Depending on the order of the fixes this would be a case for #3853.

**Comment by romainmenke** [comment]
Sure, but also no in practice :)
`no-duplicate-selectors` doesn't have autofixes.

**Comment by ybiquitous** [comment]
> Can we have a rule to enforce consistent escaping?
> Such a rule could promote avoiding needless escape sequences and promote the least verbose escape sequence.

Sounds like a good suggestion. But even if such a rule is added, I think we should fix this issue separately because each rule must be independent.

**Comment by besalraj30** [comment]
Hi

I’d like to work on this  issue. Could you assign it to me?

**Comment by ybiquitous** [comment]
@besalraj30 Thanks for your proposal. You can freely work on any issues with the "ready to implement" label without assigning.

But I want to know how to unescape selectors before implementing a fix. Do you have any plans?

**Comment by romainmenke** [comment]
See: https://github.com/postcss/postcss-selector-parser/blob/1b1e9c3bc10ccc3bc5f07a987caa7f2684c0b52f/src/selectors/className.js#L13-L28

It seems that `postcss-selector-parser` can actually normalize escaping in selectors by setting fields to their own value.

Might be a bit opaque and could be abusing `postcss-selector-parser` internals.

```js
/**
 * @param {Node} node
 */
function normalizeNodeEscaping(node) {
	// Field getters return unescaped values.
	// Raw fields contain the original escaped values.
	// Field setters handle escaping of the input and erasing the raw fields.
	// Setting any field to itself effectively normalizes CSS escaping.

	if (parser.isClassName(node)) {
		node.value = node.value.toString();
	} else if (parser.isAttribute(node)) {
		node.attribute = node.attribute.toString();
	}
}
```

Also hard to optimize because we can't know if a value needs normalization.
This would be a brute force approach. Escape everything just in case it needed escaping.

`node.raws` also doesn't seem exposed for `ClassName`.
Might just have been overlooked in the typing.

If it were exposed we could inspect `node.raws.value` to check if it contains a value with a `\`
